src/edu/csus/ecs/pc2/shadow/ShadowScoreboardComparator.java

```diff
@@ -0,0 +1,391 @@
+            
+            //get the list of TeamScoreRows out of the JSON model
+            List<TeamScoreRow> rows = jsonModel.getRows();
+            
+            //add each TeamScoreRow to the appropriate PriorityQueue in the return map, using the rank as the queue selection key
+            for (TeamScoreRow row : rows) {
+                
+                //determine the rank for the team represented by this row (note that more than one team can have the same rank, due to ties...)
+                Integer rank = row.getRank();  //uses autoboxing
+                
+                //get the current entry for this rank (if any) from the map
+                PriorityQueue<TeamScoreRow> rankQueue = retMap.get(rank);
+                
+                //add a Priority Queue for this rank if there wasn't already one there
+                if (rankQueue==null) {
+                    rankQueue = new PriorityQueue<TeamScoreRow>(new TeamScoreRowByTeamIdComparator());
+                    retMap.put(rank, rankQueue);
+                }
+                
+                rankQueue.add(row);
+            }
+            
+            //debug:
+//            System.out.println ("Scoreboard rank map:");
+//            for (Integer rank : retMap.keySet()) {
+//                PriorityQueue<TeamScoreRow> queue = retMap.get(rank);
+//                System.out.print("   Rank " + rank + ": ");
+//                for (TeamScoreRow row : queue) {
+//                    System.out.print ("  Team " + row.getTeam_id() + " Solved " + row.getScore().getNum_solved() + " Time " + row.getScore().getTotal_time()
+//                            + ";  ");
+//                }
+//                System.out.println();
+//            }
+            
+            
+        } catch (IOException e) {
+            log.severe("Exception converting scoreboard JSON string to ScoreboardJsonModel object: " + e.getMessage());
+            //debug
+//            System.out.println("Exception converting scoreboard JSON string to ScoreboardJsonModel object: " + e.getMessage());
+        }
+
+        return retMap;
+    }
+    
+
+    /**
+     * This class provides a {@link Comparator} which returns an indication of the relative ordering of two 
+     * {@link TeamScoreRow}s taking into account only the teamId.  That is, this method ranks TeamScoreRows by the numeric TeamId.
+     * 
+     * This is useful, for example, when other tests (e.g. a call to {@link TeamScoreRow#compareTo(TeamScoreRow)}
+     * has determined that two TeamScoreRows are "equal" (that is, have the same rank, number solved, and total
+     * points).  This method provides a way to order different teams in the absence of having the Team Display Name
+     * available for alphabetic ordering.
+     * 
+     * @author John Clevenger, PC2 Development Team (pc2@ecs.csus.edu)
+     */
+    private class TeamScoreRowByTeamIdComparator implements Comparator<TeamScoreRow> {
```

Comment:
fyi - this should be in its own class (and yes this is a Comparator)

Reply:
If it was in a separate file it would be accessible to other classes, and that's not the intent.
And yes, I know what a Comparator is.  I just disagree that because you call something a Comparator that it necessarily means it has to implement java.util.Comparator.   If I write a class which compares two things, it's a comparator.  If it doesn't "implement java.util.Comparator", it's not a java.util.Comparator -- but it can still be a class that compares things -- a comparator.